### ISSUE TITLE
Confuse about model SimultaneousTFWaitKNMT

### ISSUE THREAD

**Issue by xinshoulzc** [ImperialNLP/pysimt#1]
Class SimultaneousTFWaitKNMT

```Python
def forward(self, batch, **kwargs):
        """
        Performs a forward pass.
        :param batch: The batch.
        :param kwargs: Any extra arguments.
        :return: The output from the forward pass.
        """
        k = int(self.opts.model['translator_args']['k'])
        if self.training:
            epoch_count = kwargs['ectr']
            if epoch_count <= self.opts.model['consecutive_warmup']:
                # warming up, use full contexts
                k = int(1e4)

        # Pass 'k' to the model.
        return super().forward(batch, k=k)
```
In the forward process of SimultaneousTFWaitKNMT, encoders uses full context rather than 'first k token' which may cause feature leakage. I think transformer encoders should recalculate encoder hidden state in every encode step. 
Although k is passed to super class, but not used

**Comment by venetahar** [comment]
Hello,

Thank you for reaching out. I am happy to walk you through the code in a bit more details and hopefully clarify your questions. 

You are correct that the traditional transformer has a bidirectional encoder which in a simultaneous setting means that we would need to recompute these states. However, as this slows down training considerably, in the Wait-K version that we have implemented, we decided to only allow for a unidirectional encoder, as seen in the [assertion in the model](https://github.com/ImperialNLP/pysimt/blob/master/pysimt/models/snmt_tf_waitk.py#L30) which employs a lookahead mask similar to the one in the decoder self-attention layer, as done by Elbayad et al. [1], Ma et al. [2], to prevent tokens from attending to future positions.  See [this line in the TFEncoder ](https://github.com/ImperialNLP/pysimt/blob/master/pysimt/layers/encoders/transformers.py#L124). 

Regarding, passing k to the superclass, you can see that [here we pass the kwargs to the decoder](https://github.com/ImperialNLP/pysimt/blob/master/pysimt/models/snmt_tf.py#L170) which could contain the value for k. The TFDecoder uses this value and if it's present [creates another mask](https://github.com/ImperialNLP/pysimt/blob/master/pysimt/layers/decoders/tf_decoder.py#L162) to be applied in the cross attention layer which ensures that at timestep t the decoder can only attend up to the k+t−1 source tokens. We can do this only because the encoder is unidirectional and there is no information about future states encoded in each encoder state. 

I hope this makes sense. Feel free to let me know if you have any further questions. 

[1] M. Elbayad, L. Besacier, and J. Verbeek.  Efficient wait-k models for simultaneous machine translation.arXiv preprint arXiv:2005.08595, 2020. 
[2] X. Ma, J. Pino, J. Cross, L. Puzon, and J. Gu.  Monotonic multihead attention.arXiv preprint arXiv:1909.12406, 2019.

**Comment by xinshoulzc** [comment]
Thank you for quick reply. It helps me a lot. 
Comparing with unbidirectional encoder, some papers show that a bidirectional encoder significantly improves translation quality. I hope to know if you have any further plan about applying bidirectional encoders?
